crates/wasm/src/wasm_planner.rs

```diff
@@ -0,0 +1,225 @@
+            serde_wasm_bindgen::from_value(swap_commitment)?;
+
+        let swap_record = get_swap_by_commitment(swap_commitment_proto).await.unwrap();
+        let chain_params_proto: ChainParameters = get_chain_parameters().await.unwrap();
+
+        let swap_claim_plan = SwapClaimPlan {
+            swap_plaintext: swap_record.swap,
+            position: swap_record.position,
+            output_data: swap_record.output_data,
+            epoch_duration: chain_params_proto.epoch_duration,
+            proof_blinding_r: Fq::rand(&mut OsRng),
+            proof_blinding_s: Fq::rand(&mut OsRng),
+        };
+
+        self.planner.swap_claim(swap_claim_plan);
+        Ok(())
+    }
+
+    pub fn swap(
+        &mut self,
+        input_value: JsValue,
+        into_denom: JsValue,
+        swap_claim_fee: JsValue,
+        claim_address: JsValue,
+    ) -> Result<(), JsValue> {
+        let input_value_proto: Value = serde_wasm_bindgen::from_value(input_value)?;
+        let into_denom_proto: DenomMetadata = serde_wasm_bindgen::from_value(into_denom)?;
+        let swap_claim_fee_proto: Fee = serde_wasm_bindgen::from_value(swap_claim_fee)?;
+        let claim_address_proto: Address = serde_wasm_bindgen::from_value(claim_address)?;
+
+        let _ = self.planner.swap(
+            input_value_proto.try_into().unwrap(),
+            into_denom_proto.try_into().unwrap(),
+            swap_claim_fee_proto.try_into().unwrap(),
+            claim_address_proto.try_into().unwrap(),
+        );
+
+        Ok(())
+    }
+
+    pub async fn plan(&mut self, self_address: JsValue) -> Result<JsValue, JsValue> {
+        utils::set_panic_hook();
+
+        let self_address_proto: Address = serde_wasm_bindgen::from_value(self_address)?;
+
+        let chain_params_proto: ChainParameters = get_chain_parameters().await.unwrap();
+        let fmd_params_proto: FmdParameters = get_fmd_parameters().await.unwrap();
+
+        let mut spendable_notes = Vec::new();
+
+        let (spendable_requests, _) = self.planner.notes_requests();
+
+        for request in spendable_requests {
+            let notes = get_notes(request);
+            spendable_notes.extend(notes.await.unwrap());
+        }
+
+        // Plan the transaction using the gathered information
+
+        let plan: penumbra_transaction::plan::TransactionPlan = self
+            .planner
+            .plan_with_spendable_and_votable_notes(
+                &chain_params_proto.try_into().unwrap(),
+                &fmd_params_proto.try_into().unwrap(),
+                spendable_notes,
+                Vec::new(),
+                self_address_proto.try_into().unwrap(),
+            )
+            .unwrap();
+
+        let plan_proto: TransactionPlan = plan.to_proto();
+
+        Ok(serde_wasm_bindgen::to_value(&plan_proto)?)
+    }
+}
+
+pub async fn get_chain_parameters() -> Option<ChainParameters> {
+    let db_req: OpenDbRequest = IdbDatabase::open_u32("penumbra", 12).ok()?;
```

Comment:
I think we should have a `IndexDb` struct that has all of these helper functions in one place in another file.